// matrixjs-tx/es5/index.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ethUtil = require('matrixjs-util');
var fees = require('ethereum-common/params.json');
var BigNumber = require('bignumber.js')
const SimpleBase = require('simple-base')
var BN = ethUtil.BN;

// secp256k1n/2
var N_DIV_2 = new BN('7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0', 16);

/**
 * Creates a new transaction object.
 *
 * @example
 * var rawTx = {
 *   nonce: '00',
 *   gasPrice: '09184e72a000',
 *   gasLimit: '2710',
 *   to: '0000000000000000000000000000000000000000',
 *   value: '00',
 *   data: '7f7465737432000000000000000000000000000000000000000000000000000000600057',
 *   v: '1c',
 *   r: '5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
 *   s: '5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
 * };
 * var tx = new Transaction(rawTx);
 *
 * @class
 * @param {Buffer | Array | Object} data a transaction can be initiailized with either a buffer containing the RLP serialized transaction or an array of buffers relating to each of the tx Properties, listed in order below in the exmple.
 *
 * Or lastly an Object containing the Properties of the transaction like in the Usage example.
 *
 * For Object and Arrays each of the elements can either be a Buffer, a hex-prefixed (0x) String , Number, or an object with a toBuffer method such as Bignum
 *
 * @property {Buffer} raw The raw rlp encoded transaction
 * @param {Buffer} data.nonce nonce number
 * @param {Buffer} data.gasLimit transaction gas limit
 * @param {Buffer} data.gasPrice transaction gas price
 * @param {Buffer} data.to to the to address
 * @param {Buffer} data.value the amount of ether sent
 * @param {Buffer} data.data this will contain the data of the message or the init of a contract
 * @param {Buffer} data.v EC signature parameter
 * @param {Buffer} data.r EC signature parameter
 * @param {Buffer} data.s EC recovery ID
 * @param {Number} data.chainId EIP 155 chainId - mainnet: 1, ropsten: 3
 * */

var Transaction = function () {
  function Transaction(data, mutilTrans = true) {
    _classCallCheck(this, Transaction, mutilTrans);

    data = data || {};
    // Define Properties
    var mutil_fields_ = [{
      name: 'nonce',
      length: 32,
      allowLess: true,
      index: 0,
      default: new Buffer([])
    }, {
      name: 'gasPrice',
      length: 32,
      allowLess: true,
      index: 1,
      default: new Buffer([])
    }, {
      name: 'gasLimit',
      alias: 'gas',
      length: 32,
      allowLess: true,
      index: 2,
      default: new Buffer([])
    }, {
      name: 'to',
      allowZero: true,
      index: 3,
      default: ''
    }, {
      name: 'value',
      length: 32,
      allowLess: true,
      index: 4,
      default: new Buffer([])
    }, {
      name: 'data',
      alias: 'input',
      allowZero: true,
      index: 5,
      default: new Buffer([])
    }, {
      name: 'v',
      allowZero: true,
      index: 6,					// 7
      default: new Buffer([0x1c])
    }, {
      name: 'r',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 7,					// 8
      default: new Buffer([])
    }, {
      name: 's',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 8,					// 9
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'TxEnterType',
      allowZero: true,
      allowLess: true,
      index: 9,				// 6
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'IsEntrustTx',
      allowZero: true,
      allowLess: true,
      index: 10,				// 6
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'CommitTime',
      allowZero: true,
      allowLess: true,
      index: 11,				// 6
      default: new Buffer([])
    }, {
      name: 'extra_to',
      allowZero: true,
      allowLess: true,
      index: 12,				// 6
      default: new Buffer([])
    }];

    var mutil_fields = [{
      name: 'nonce',
      length: 32,
      allowLess: true,
      index: 0,
      default: new Buffer([])
    }, {
      name: 'gasPrice',
      length: 32,
      allowLess: true,
      index: 1,
      default: new Buffer([])
    }, {
      name: 'gasLimit',
      alias: 'gas',
      length: 32,
      allowLess: true,
      index: 2,
      default: new Buffer([])
    }, {
      name: 'to',
      allowZero: true,
      index: 3,
      default: ''
    }, {
      name: 'value',
      length: 32,
      allowLess: true,
      index: 4,
      default: new Buffer([])
    }, {
      name: 'data',
      alias: 'input',
      allowZero: true,
      index: 5,
      default: new Buffer([])
    }, {
      name: 'v',
      allowZero: true,
      index: 6,					// 7
      default: new Buffer([0x1c])
    }, {
      name: 'r',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 7,					// 8
      default: new Buffer([])
    }, {
      name: 's',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 8,					// 9
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'TxEnterType',
      allowZero: true,
      allowLess: true,
      index: 9,				// 6
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'IsEntrustTx',
      allowZero: true,
      allowLess: true,
      index: 10,				// 6
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'CommitTime',
      allowZero: true,
      allowLess: true,
      index: 11,				// 6
      default: new Buffer([])
    }, {                         // 特殊处理
      name: 'extra_to',
      allowZero: true,
      allowLess: true,
      index: 12,				// 6
      default: new Buffer([])
    }];

    var single_fields = [{
      name: 'nonce',
      length: 32,
      allowLess: true,
      index: 0,
      default: new Buffer([])
    }, {
      name: 'gasPrice',
      length: 32,
      allowLess: true,
      index: 1,
      default: new Buffer([])
    }, {
      name: 'gasLimit',
      alias: 'gas',
      length: 32,
      allowLess: true,
      index: 2,
      default: new Buffer([])
    }, {
      name: 'to',
      allowZero: true,
      index: 3,
      default: ''
    }, {
      name: 'value',
      length: 32,
      allowLess: true,
      index: 4,
      default: new Buffer([])
    }, {
      name: 'data',
      alias: 'input',
      allowZero: true,
      index: 5,
      default: new Buffer([])
    }, {
      name: 'v',
      allowZero: true,
      index: 6,
      default: new Buffer([0x1c])
    }, {
      name: 'r',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 7,
      default: new Buffer([])
    }, {
      name: 's',
      length: 32,
      allowZero: true,
      allowLess: true,
      index: 8,
      default: new Buffer([])
    }];


      var mutil_fields = [{
          name: 'nonce',
          length: 32,
          allowLess: true,
          index: 0,
          default: new Buffer([])
      }, {
          name: 'gasPrice',
          length: 32,
          allowLess: true,
          index: 1,
          default: new Buffer([])
      }, {
          name: 'gasLimit',
          alias: 'gas',
          length: 32,
          allowLess: true,
          index: 2,
          default: new Buffer([])
      }, {
          name: 'to',
          allowZero: true,
          index: 3,
          default: ''
      }, {
          name: 'value',
          length: 32,
          allowLess: true,
          index: 4,
          default: new Buffer([])
      }, {
          name: 'data',
          alias: 'input',
          allowZero: true,
          index: 5,
          default: new Buffer([])
      }, {
          name: 'v',
          allowZero: true,
          index: 6,					// 7
          default: new Buffer([0x1c])
      }, {
          name: 'r',
          length: 32,
          allowZero: true,
          allowLess: true,
          index: 7,					// 8
          default: new Buffer([])
      }, {
          name: 's',
          length: 32,
          allowZero: true,
          allowLess: true,
          index: 8,					// 9
          default: new Buffer([])
      }, {                         // 特殊处理
          name: 'TxEnterType',
          allowZero: true,
          allowLess: true,
          index: 9,				// 6
          default: new Buffer([])
      }, {                         // 特殊处理
          name: 'IsEntrustTx',
          allowZero: true,
          allowLess: true,
          index: 10,				// 6
          default: new Buffer([])
      }, {                         // 特殊处理
          name: 'CommitTime',
          allowZero: true,
          allowLess: true,
          index: 11,				// 6
          default: new Buffer([])
      }, {                         // 特殊处理
          name: 'extra_to',
          allowZero: true,
          allowLess: true,
          index: 12,				// 6
          default: new Buffer([])
      }];

      var single_fields = [{
          name: 'v',
          allowZero: true,
          index: 1,
          default: new Buffer([0x1c])
      }, {
          name: 'r',
          length: 32,
          allowZero: true,
          allowLess: true,
          index: 2,
          default: new Buffer([])
      }, {
          name: 's',
          length: 32,
          allowZero: true,
          allowLess: true,
          index: 3,
          default: new Buffer([])
      }];

    /**
     * Returns the rlp encoding of the transaction
     * @method serialize
     * @return {Buffer}
     * @memberof Transaction
     * @name serialize
     */
    // attached serialize

    if (mutilTrans){
      if ((typeof data === 'undefined' ? 'undefined' : typeof data) === 'object') {
        ethUtil.defineProperties(this, mutil_fields_, data);
      } else if (typeof data === 'string') {
        ethUtil.defineProperties(this, mutil_fields, data);
      }
    } else {
      ethUtil.defineProperties(this, single_fields, data);
    }

    /**
     * @property {Buffer} from (read only) sender address of this transaction, mathematically derived from other parameters.
     * @name from
     * @memberof Transaction
     */
    Object.defineProperty(this, 'from', {
      enumerable: true,
      configurable: true,
      get: this.getSenderAddress.bind(this)
    });

    // calculate chainId from signature
    var sigV = ethUtil.bufferToInt(this.v);
    // if (typeof this.extra_to != 'undefined' && this.extra_to.length > 0) {
    //     sigV -= 128;
    // }
    var chainId = Math.floor((sigV - 35) / 2);
    if (chainId < 0) chainId = 0;

    // set chainId
    this._chainId = chainId || data.chainId || 0;
    this._homestead = true;
  }

  /**
   * If the tx's `to` is to the creation address
   * @return {Boolean}
   */


  Transaction.prototype.toCreationAddress = function toCreationAddress() {
    return this.to.toString('hex') === '';
  };

  /**
   * Computes a sha3-256 hash of the serialized tx
   * @param {Boolean} [includeSignature=true] whether or not to inculde the signature
   * @return {Buffer}
   */


  Transaction.prototype.hash = function hash(includeSignature, trans) {
    if (includeSignature === undefined) includeSignature = true;

    // EIP155 spec:
    // when computing the hash of a transaction for purposes of signing or recovering,
    // instead of hashing only the first six elements (ie. nonce, gasprice, startgas, to, value, data),
    // hash nine elements, with v replaced by CHAIN_ID, r = 0 and s = 0

    var items = void 0;
    if (includeSignature) {
      items = this.raw;
    } else {
      if (this._chainId > 0) {
        var raw = this.raw.slice();
        this.v = this._chainId;
        this.r = 0;
        this.s = 0;
        items = this.raw;
        this.raw = raw;
      } else {
        items = this.raw.slice(0, 6);
      }
    }
    if (trans){
      var raws = [];
      this._extra_fields.forEach(function (field) {
        raws.push(items[field.index]);
      });
    } else {
      raws = items
    }
    // create hash
    console.log("+++++++ \n", "trans: ", trans, raws);
    console.log(ethUtil.rlphash(raws).toString('hex'))
    return ethUtil.rlphash(raws);
  };

  /**
   * returns the public key of the sender
   * @return {Buffer}
   */


  Transaction.prototype.getChainId = function getChainId() {
    return this._chainId;
  };

  /**
   * returns the sender's address
   * @return {Buffer}
   */


  Transaction.prototype.getSenderAddress = function getSenderAddress() {
    if (this._from) {
      return this._from;
    }
    var pubkey = this.getSenderPublicKey();
    this._from = ethUtil.publicToAddress(pubkey);
    console.log(this._from.toString('hex'))
    return this._from;
  };
  /**
   * returns the public key of the sender
   * @return {Buffer}
   */


  Transaction.prototype.getSenderPublicKey = function getSenderPublicKey() {
    if (!this._senderPubKey || !this._senderPubKey.length) {
      if (!this.verifySignature()) throw new Error('Invalid Signature');
    }
    return this._senderPubKey;
  };

  /**
   * Determines if the signature is valid
   * @return {Boolean}
   */


  Transaction.prototype.verifySignature = function verifySignature() {
    var msgHash = this.hash(false, true);
    console.log("verifySignature: \n", "trans: ", true, msgHash);
    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
    if (this._homestead && new BN(this.s).cmp(N_DIV_2) === 1) {
      return false;
    }
    try {
      var v = ethUtil.bufferToInt(this.v);
      if (this._chainId > 0) {
        v -= this._chainId * 2 + 8;
        // if (typeof this.extra_to != 'undefined' && this.extra_to.length > 0) {
        //     v -= 128;
        // }
      }
      this._senderPubKey = ethUtil.ecrecover(msgHash, v, this.r, this.s);
    } catch (e) {
      return false;
    }

    return !!this._senderPubKey;
  };

  /**
   * sign a transaction with a given a private key
   * @param {Buffer} privateKey
   */


  Transaction.prototype.sign = function sign(privateKey) {
    var msgHash = this.hash(false, false);
    console.log("sign: \n", "trans: ", false, msgHash);
    var sig = ethUtil.ecsign(msgHash, privateKey);
    if (this._chainId > 0) {
      sig.v += this._chainId * 2 + 8;
      // if (typeof this.extra_to != 'undefined' && this.extra_to.length > 0){
      //     sig.v += 128;
      // }
    }
    Object.assign(this, sig);
  };

  /**
   * The amount of gas paid for the data in this tx
   * @return {BN}
   */


  Transaction.prototype.getDataFee = function getDataFee() {
    var data = this.raw[5];
    var cost = new BN(0);
    for (var i = 0; i < data.length; i++) {
      data[i] === 0 ? cost.iaddn(fees.txDataZeroGas.v) : cost.iaddn(fees.txDataNonZeroGas.v);
    }
    return cost;
  };

  /**
   * the minimum amount of gas the tx must have (DataFee + TxFee + Creation Fee)
   * @return {BN}
   */


  Transaction.prototype.getBaseFee = function getBaseFee() {
    var fee = this.getDataFee().iaddn(fees.txGas.v);
    if (this._homestead && this.toCreationAddress()) {
      fee.iaddn(fees.txCreation.v);
    }
    return fee;
  };

  /**
   * the up front amount that an account must have for this transaction to be valid
   * @return {BN}
   */


  Transaction.prototype.getUpfrontCost = function getUpfrontCost() {
    return new BN(this.gasLimit).imul(new BN(this.gasPrice)).iadd(new BN(this.value));
  };

  /**
   * validates the signature and checks to see if it has enough gas
   * @param {Boolean} [stringError=false] whether to return a string with a dscription of why the validation failed or return a Bloolean
   * @return {Boolean|String}
   */


  Transaction.prototype.validate = function validate(stringError) {
    var errors = [];
    if (!this.verifySignature()) {
      errors.push('Invalid Signature');
    }

    if (this.getBaseFee().cmp(new BN(this.gasLimit)) > 0) {
      errors.push(['gas limit is too low. Need at least ' + this.getBaseFee()]);
    }

    if (stringError === undefined || stringError === false) {
      return errors.length === 0;
    } else {
      return errors.join(' ');
    }
  };

  Transaction.prototype.getTxParams = function getTxParams(hash) {
    let txData = new Transaction(hash, true)
    let newTxData = {
      v: '',
      r: '',
      s: '',
      data: '',
      gasPrice: '',
      gas: '',
      value: '',
      nonce: '',
      currency: '',
      txType: 0,
      lockHeight: 0,
      isEntrustTx: 0,
      txEnterType: 0,
      commitTime: 0,
      extra_to: []
    }
    newTxData.v = '0x' + txData.v.toString('hex')
    newTxData.r = '0x' + txData.r.toString('hex')
    newTxData.s = '0x' + txData.s.toString('hex')
    newTxData.data = '0x' + txData.data.toString('hex')
    newTxData.gasPrice = '0x' + txData.gasPrice.toString('hex')
    newTxData.gas = '0x' + txData.gasLimit.toString('hex')
    newTxData.value = '0x' + txData.value.toString('hex')
    if (newTxData.value === '0x') {
      newTxData.value = '0x0'
    }
    newTxData.nonce = '0x' + txData.nonce.toString('hex')
    if (txData.to.toString() !== '') {
      newTxData.to = txData.to.toString()
      newTxData.currency = txData.to.toString().split('.')[0]
    } else {
      newTxData.currency = 'MAN'
    }
    let isEntrustTx = '0x' + txData.IsEntrustTx.toString('hex')
    let txEnterType = '0x' + txData.TxEnterType.toString('hex')
    let commitTime = '0x' + txData.CommitTime.toString('hex')
    let txType = '0x' + txData.extra_to[0][0].toString('hex')
    let lockHeight = '0x' + txData.extra_to[0][1].toString('hex')
    newTxData.txType = txType === '0x' ? 0 : Number(this.hexToDecimal(txType))
    newTxData.lockHeight = lockHeight === '0x' ? 0 : Number(this.hexToDecimal(lockHeight))
    newTxData.isEntrustTx = isEntrustTx === '0x' ? 0 : Number(this.hexToDecimal(isEntrustTx))
    newTxData.txEnterType = txEnterType === '0x' ? 0 : Number(this.hexToDecimal(txEnterType))
    newTxData.commitTime = commitTime === '0x' ? 0 : Number(this.hexToDecimal(commitTime))
    newTxData.extra_to = []
    for (var i = 0, length = txData.extra_to[0][2].length; i < length; i++) {
      newTxData.extra_to.push({
        to: txData.extra_to[0][2][i][0].toString(),
        value: '0x' + txData.extra_to[0][2][i][1].toString('hex')
      })
    }
    return newTxData;
  };
  Transaction.prototype.hexToDecimal = function hexToDecimal(hex) {
    return new BigNumber(this.sanitizeHex(hex)).toString()
  }
  Transaction.prototype.sanitizeHex = function sanitizeHex(hex) {
    hex = hex.substring(0, 2) === '0x' ? hex.substring(2) : hex
    if (hex === '') return ''
    return '0x' + this.padLeftEven(hex)
  }
  Transaction.prototype.padLeftEven = function padLeftEven(hex) {
    hex = hex.length % 2 !== 0 ? '0' + hex : hex
    return hex
  }

  return Transaction;

}();

module.exports = Transaction;
